README.md
# Saleor Storefront

![1 copy 2x](https://user-images.githubusercontent.com/5421321/47798207-30aeea00-dd28-11e8-9398-3d8426836a83.png)

_**Note:** This project is beta quality. We don't advise using it in production._

A GraphQL-powered, PWA, single-page application storefront for [Saleor](https://github.com/mirumee/saleor/).

## Features

- Headless ecommerce storefront built with [GraphQL](https://graphql.org/), [Apollo Client](https://www.apollographql.com/client), [React](https://reactjs.org/) and [Typescript](https://www.typescriptlang.org/)
- Offline mode (beta)
- Saleor GraphQL API integration
- Single-page application experience
- [Braintree Payment Gateway](https://www.braintreepayments.com/) integration

## Demo

See the [public demo](http://demo.saleor.io) of Saleor Storefront!

Or launch the demo on a free Heroku instance.

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy)

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.

### Prerequisites

- Node.js 10.0+
- A running instance of Saleor.

  To run the storefront, you have to set the `API_URI` environment variable to point to the Saleor GraphQL API. If you are running Saleor locally with the default settings, set `API_URI` to: `http://localhost:8000/graphql/`.

### Installing

Clone the repository:

```
git clone https://github.com/mirumee/saleor-storefront.git
```

Enter the project directory:

```
cd saleor-storefront
```

#### Using stable release

To use the official stable release, checkout to a release tag:

```
$ git checkout 2.10.4
```

See the list of all releases here: https://github.com/mirumee/saleor-storefront/releases/

#### Using development version

If you want to use the latest development version, checkout to the `master` branch:

```
$ git checkout master
```

Install NPM dependencies:

```
npm i
```

Run the development server:

```
npm start
```

Go to `http://localhost:3000` to access the storefront.

## Cypress tests

If you want to run [Cypress](https://www.cypress.io/) tests, make sure that all dependecies (including `Cypress`) are installed by running the install command.

```
npm i
```

Following environment variables are required to be set in order to be able to run tests properly:

- `API_URI` - GraphQL API address.
- `STATIC_URL` - static files destination url, eg. S3 bucket
- `CYPRESS_USER_NAME` - username (email) for `Storefront` user.
- `CYPRESS_USER_PASSWORD` - for the user mentioned above.

If you are runninng the Storefront from the perspective of `Docker` container, then you can run tests using following commands:

Headless mode:

```
cy:run
```

Cypress UI mode:

```
cy:open
```

If you want to run tests against your local development environment then use following commands:

Headless mode:

```
test:e2e:run
```

Cypress UI mode:

```
test:e2e:dev
```

## Creating new components

All new components should follow Atomic Design Guidelines and be placed in `src/@next/components` directory.

Files structure can be generated using `plop`:

```
npm run generate
```

## Modifying the Storefront

[From Spectrum Post](https://spectrum.chat/saleor/saleor-storefront/modifying-the-storefront~c1955dbf-a421-4fb6-b99e-937dd2642b23)

### Important Files

- **saleor-storefront/config/webpack/config.base.js** - Base webpack config file.
  - Can change name of the app (displayed when installed on mobile)
- **saleor-storefront/src/index.html** - Main template file that contains the <div id="root"></div>
  - Can change title of storefront here
- **saleor-storefront/src/index.tsx** - Main entry point file. Render's the <App /> component, apollo-client, and others to the root div in index.html file above.
- **saleor-storefront/src/core/config.ts** - Controls number of products shown per page, support email, gateway providers, social media, and some meta.
  - Can change support email
  - Can change products shown per page
  - Can change gateway providers
  - Can change social media links that are displayed in the footer
  - Can change some meta options
- **saleor-storefront/src/images/** - Holds all the images for logo, cart, favicon, etc.
  - Can change storefront logo, favicon, or add new images here.
- **saleor-storefront/src/globalStyles/scss/variables.scss** - Contains base styles like colors, font size, container width, media breakpoints and more.
- **saleor-storefront/src/@next/globalStyles/** - Contains more base styles, themes, media, and constants.
- **saleor-storefront/src/views/** - This folder controls the views, or what is displayed for each page. Most views have a file named "Page.tsx" that controls the layout of the page and a file named "View.tsx" that calls the query and renders the <Page /> component with the data.
  - Can add another view to storefront here. Requires adding a route (see routes below).
- saleor-storefront/src/@next/pages/ - Second spot for modifying/adding different pages. This is the recommended directory to add new pages.
- **saleor-storefront/src/app/routes/** - This folder contains the paths as well as holds the <Routes /> component. Here is where you add a new path and route.
  1.  Export a new path in paths.ts
  2.  Inside AppRoutes.tsx import your new view (see views above) and create a new route with path={paths.newPath} and component={newViewPage}
  3.  To link to your new view import { Link } from "react-router-dom" and use new path you created in paths.ts (make sure to import it)
- **saleor-storefront/src/app/App.tsx** - This is main <App /> component that renders the <MainMenu />, <Routes /> (explained below), <Footer /> and a couple other components.

### Adding a Payment Gateway

- **saleor-storefront/src/core/config.ts** - Add new gateway provider name here.
- **saleor-storefront/src/@next/components/organisms/** - Create a new folder for new payment gateway component here.
- **saleor-storefront/src/@next/components/organisms/PaymentGatewaysList/PaymentGatewaysList.tsx** - Import new gateway component, create a new switch case to handle your gateway component.

### Receiving confirmation emails

- **Set [EMAIL_URL](https://docs.saleor.io/docs/developer/running-saleor/configuration#setting-environment-variables) environment variable for Saleor core.**
  - Using Docker - Add EMAIL_URL as new enviornment variable to both the api and worker service following the format [here](https://docs.saleor.io/docs/developer/running-saleor/configuration#email_url).
- **Issues getting emails working?**
  - Gmail
    - Check to see that "Less secure app access" is turned ON. Under "Manage your Google Account" > Go to the security tab. By default, the setting is off for security reasons.
    - If using 2FA make sure to set an [app password](https://support.google.com/accounts/answer/185833?p=InvalidSecondFactor&visit_id=637355441414497566-1310044707&rd=1) and use that in place of your normal login password.

### Multichannel

- **Set [DEFAULT_CHANNEL_SLUG] environment variable.**

## License

This project is licensed under the BSD-3-Clause License - see the [LICENSE](https://github.com/mirumee/saleor-storefront/blob/master/LICENSE) file for details

#### Crafted with ❤️ by [Mirumee Software](http://mirumee.com)

hello@mirumee.com


CHANGELOG.md
# Changelog

All notable, unreleased changes to this project will be documented in this file. For the released changes, please visit the [Releases](https://github.com/mirumee/saleor-storefront/releases) page.

## [Unreleased]

## 2.11.0

- Add product hyperlink in cart page - #745 by @konstantinoschristomanos
- Add Cypress tags to all of the buttons (also speed up tests) - #718 by @krzysztofwolski
- Automatically choose variant attributes from url in ProductVariantPicker - #708 by @AlicjaSzu
- Use SDK as a standalone package - #724 by @dominik-zeglen
- Use auth API from SDK - #727 by @orzechdev
- Fix `CartRow` tests - #749 by @dominik-zeglen
- Add prettier to precommit - #766 by @dominik-zeglen
- Do not use prettier on generated files - #773 by @dominik-zeglen
- Add Eslint - #776 by @dominik-zeglen
- Fix storybook config - #789 by @dominik-zeglen
- Update SDK to newest version - #795 by @dominik-zeglen
- Use intl - #782 by @przlada
- Download invoice for order - #790 by @orzechdev
- Do not throw error if unsupported payment gateway found - #819 by @dominik-zeglen
- Fix tsconfig aliases - #824 by @orzechdev
- Set billing address in first checkout step - #822 by @orzechdev
- Persist payment gateways for the whole checkout - #828 by @orzechdev
- Add test tags to address book - #847 by @dominik-zeglen
- Refresh user data if mutation fails - #854 by @dominik-zeglen
- Fix product list data overfetching - #855 by @dominik-zeglen
- Add Adyen payment gateway - #845 by @orzechdev
- Fix crash on null price range - #875 by @orzechdev
- Update collection products query - #879 by @orzechdev
- Fix checkout refreshing - #865 by @orzechdev
- Add purchase availability to product details page - #878 by @orzechdev
- Require payment recreate when payment price is wrong - #892 by @orzechdev
- Handle JWT token refreshing and verifying - #883 by @orzechdev
- Fix cart sidebar style - #897 by @orzechdev
- Refactor variant picker components and open sidebar after adding product to cart - #809 by @krzysztofwolski
- Fix checkout address view - #909 by @konstantinoschristomanos
- Support for static URL - #721 by @marianoeramirez and @dominik-zeglen
- Fix search crashing when displaying item with no category - #928 by @mmarkusik
- Fix generating site map - #915 by @rboixaderg
- Use defaultChannelSlug in queries and mutations - #937 by @AlicjaSzu

## 2.10.4

- Fix build errors introduced in version 2.10.3 - by @dominik-zeglen

## 2.10.3

- Stop storing plain text passwords in localStorage - by @dominik-zeglen

## 2.10.2

- Fix fetching `quantityAvailable` field - #738 by @AlicjaSzu

## 2.10.1

- Replace stockQuantity field with quantityAvailable - #723 by @AlicjaSzu
- Regenerate types - #712 by @dominik-zeglen

## 2.10.0

- Account confirmation mechanism - #565 by @tomaszszymanski129
- Add missing product attributes on product page - #536 by @orzechdev
- Change register mutation to accountRegister - #549 by @tomaszszymanski129
- Add `ProductVariantPicker` component supporting multiple product variant attributes - #550 by @orzechdev
- Fix not working storefront when no data in saleor database exist - #551 by @orzechdev
- Make checkout working without shipping if it is not required - #571 by @orzechdev
- Add ability to apply a promo code in checkout - #582 by @orzechdev
- Refactor product list - #591 by @orzechdev
- Add chips of selected filter attributes values and fix product list filtering - #602 by @orzechdev
- Refactor FilterAttribute to AttributeValuesChecklist - #610 by @orzechdev
- Clear cache on logout - #623 by @orzechdev
- Add missing Cypress functional tests - #624 by @mateuszkula
- Add lighthouse config - #627 by @mateuszkula
- Load Stripe asynchronously - #629 by @orzechdev
- Fix crash address book on logout - #630 by @orzechdev
- Preload css for Inter font - #631 by @orzechdev
- Use sdk for fetching shop details - #632 by @mateuszkula
- Preconnect to Graphql API_URI - #634 by @mateuszkula
- Upgrade typescript to 3.8.2 - #635 by @mateuszkula
- Remove linguijs usage - #637 by @mateuszkula
- Remove old storybook - #638 by @mateuszkula
- Add docs to storybook - #614 by @orzechdev
- Create new UI for product page - #605 by @mateuszkula
- Use new pricing on product page with TaxedMoney component to display prices - #584 by @orzechdev
- Update GraphQL schema - #567 by @orzechdev
- Completely recreate checkout and cart, with new checkout nad cart SDK - #639 by @orzechdev, @mateuszkula
- Hide payment options - #678 by @orzechdev
- Fix unhandled JWT token expiration - #696 by @orzechdev
- Regenerate types - #712 by @dominik-zeglen
- Replace stockQuantity field with quantityAvailable - #723 by @AlicjaSzu

## 0.7.0

- Fix login and registration overlay not showing - #322 by @mateuszkula
- Add new design for 404 page - #183 by @mateuszkula
- Add Sitemap generator - #342 by @bogdal
- Add Cypress tests - #333 by @AlicjaSzu
- Add rich-text content renderer - #361 by @AlicjaSzu
- Add rich-text content renderer for pages - #426 @ChanceLeachman
- Add TextField and ErrorMessage components - #373 by @AlicjaSzu
- Add CreditCardForm component - #369 by @AlicjaSzu
- Display filters when no product was found - #319 by @aldomonteiro
- Add ServiceWorker provider - #352 by @bogdal
- Add ButtonLink atom component - #392 by @AlicjaSzu
- Add `lingui` - #382 by @AlicjaSzu
- Add FormFooter molecule component - #393 by @AlicjaSzu
- Add Overlay component - #402 by @AlicjaSzu
- Add Modal component - #391 by @AlicjaSzu
- Add the Credential Management API support - #409 by @bogdal
- Add size picker component - #425 by @bogdal
- Replace `BACKEND_URL` in favor of `API_URI` - #474 by @bogdal
- Adapt `checkout.availablePaymentGateways` structure to the new schema - #483 by @bogdal
- Replace product/variant `price` fields with appropriate `pricing` fields - #483 by @bogdal
- Add Stripe integration - #486 by @bogdal
- Add `PlaceholderImage` component and show it in case of missing thumbnail - #489 by @xit4

## 0.6.0


config/webpack/config.base.js
const HtmlWebpackPlugin = require("html-webpack-plugin");
const WebappWebpackPlugin = require("webapp-webpack-plugin");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const TsconfigPathsPlugin = require("tsconfig-paths-webpack-plugin");
const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");
const webpack = require("webpack");
const path = require("path");

if (!process.env.API_URI) {
  throw new Error("Environment variable API_URI not set");
}
const STATIC_URL = process.env.STATIC_URL || "/";

module.exports = ({ sourceDir, distDir }) => ({
  devtool: "source-map",
  entry: {
    app: `${sourceDir}/index.tsx`,
  },
  module: {
    rules: [
      {
        exclude: /node_modules/,
        loader: "ts-loader",
        options: {
          experimentalWatchApi: true,
          transpileOnly: true,
        },
        test: /\.tsx?$/,
      },
      {
        exclude: /node_modules/,
        loader: "babel-loader",
        options: {
          configFile: "./babel.config.js",
        },
        test: /\.(jsx?|tsx?)$/,
      },
      {
        test: /\.(woff2?|ttf|eot)$/,
        use: [
          {
            loader: "file-loader",
            options: {
              name: "[name].[ext]",
              outputPath: "fonts/",
              publicPath: "/fonts/",
            },
          },
        ],
      },
      {
        test: /\.(gif|jpg|png|svg)$/,
        use: [
          {
            loader: "file-loader",
            options: {
              name: "[name].[ext]",
              outputPath: "images/",
              publicPath: "/images/",
            },
          },
          {
            loader: "image-webpack-loader",
            options: {
              gifsicle: {
                enabled: false,
              },
              mozjpeg: {
                progressive: true,
                quality: 85,
              },
              pngquant: {
                quality: "65-90",
                speed: 4,
              },
            },
          },
        ],
      },
    ],
  },
  node: {
    fs: "empty",
    module: "empty",
  },
  output: {
    path: distDir,
    publicPath: STATIC_URL,
  },
  plugins: [
    new CleanWebpackPlugin({
      cleanOnceBeforeBuildPatterns: distDir,
    }),
    new HtmlWebpackPlugin({
      API_URI: process.env.API_URI,
      filename: `${distDir}/index.html`,
      template: `${sourceDir}/index.html`,
    }),
    new ForkTsCheckerWebpackPlugin({
      eslint: true,
      exclude: "node_modules",
    }),
    // PWA plugins
    new WebappWebpackPlugin({
      favicons: {
        appDescription: "Storefront for the Saleor e-commerce platform",
        appName: "Saleor",
        background: "#ddd",
        developerURL: null, // prevent retrieving from the nearest package.json
        display: "standalone",
        theme_color: "#333",
      },
      logo: `${sourceDir}/images/favicon.png`,
      prefix: "images/favicons/",
    }),
    new webpack.EnvironmentPlugin({
      API_URI: "http://localhost:8000/graphql/",
      DEFAULT_CHANNEL_SLUG: "default-channel",
      DEMO_MODE: false,
      GTM_ID: undefined,
      SENTRY_APM: "0",
      SENTRY_DSN: null,
    }),
  ],
  resolve: {
    alias: {
      // Explicitely set react's path here because npm-link doesn't do well
      // when it comes to peer dependencies, and we need to somehow develop
      // @saleor/sdk package
      react: path.resolve("./node_modules/react"),
      "react-dom": "@hot-loader/react-dom",
    },
    extensions: [".ts", ".tsx", ".js", ".jsx"],
    plugins: [
      new TsconfigPathsPlugin({
        configFile: "./tsconfig.json",
      }),
    ],
  },
});
